scipio/src/sys/uring.rs

```diff
@@ -891,10 +872,10 @@ impl Reactor {
         let mut should_sleep = match timeout {
             None => true,
             Some(dur) => {
-                let mut src = self.timeout_src.borrow_mut();
-                lat_ring.cancel_preempt_timer(&mut src);
+                // Cancel the old timer.
+                self.timeout_src.take();
```

Comment:
Does Rust guarantee that the timer will be dropped at exactly this point and not later?

For correctness the timer must be gone before we call `flush_cancellations` below. If we don't do that we may end up racing and cancelling the new timer we just installed (if they use the same id), or having a spurious wake up if they don't.

My biggest concern when I started reading your patch was "what about flushing the cancellations?" 
But you kept it in a good place. If we cancel here then I think your change is good.

Reply:
Yes, rust does guarantee this. We can over-emphasize this by writing `drop(self.timeout_src.take())`, but I am not a fan of spelling `drop` just for emphasis.  Couple of related things:

* Drop order in Rust is not influenced by lifetimes. The (implicit) drops are inserted before the  borrow checker is run, the checker then just verifies that it is possible to work-out reasonable lifetimes with the given drops.
* `let _ = foo();` will drop the result of `foo()` immediately, `let _foo = foo();`  will keep it around until the end of the block. 
* I am not a fan of explicit `drop`, because it may change the drop order, but only on the happy path:

```rust
let a = foo();
let b = bar()

spam()?;

drop(a)
drop(b)
```

In this case, the drop order would be `a, b`  if `spam` returns Ok and `b, a` if it Errs. I've once hit a deadlock on program termination for exactly this reason. Since then, I just use extra `{}` and order local variables such that implicit drop order is the correct one.